### DISCUSSION TITLE
How to use self-hosted minio instance

### DISCUSSION THREAD

**Discussion by ivanjx** [Genbox/SimpleS3#57] [General]
Hi

I can see that this library supports minio on the providers page but I cant find the guide or sample code of how to do it. Any help would be appreciated.

**Comment by Genbox** [comment]
Minio tries to replicate Amazon S3, so you can simply use the library the same way. [See the example code here](https://github.com/Genbox/SimpleS3/blob/master/Src/SimpleS3.Examples/Program.cs).

**Comment by ivanjx** [comment]
thank you for the response @Genbox 

unfortunately i am still unable to find an example of how to set the endpoint to my minio server from your example. i am also planning to use this to connect to digital ocean's spaces which also has custom endpoint but implements S3 protocol.

**Reply by Genbox** [reply]
There was supposed to be a generic client that could connect to an arbitrary endpoint, but I seem to have misplaced it. Woops.

I'll get right on adding it back in.

**Reply by Genbox** [reply]
Note that you can still use the AmazonS3Client against minio. You can specify the endpoint you want to connect to in the AmazonS3Config. For example:

```
AmazonS3Config config = new AmazonS3Config();
config.Credentials = new StringAccessKey("YourKeyId", "YourSecretKey");
config.Endpoint = new Uri("https://yourminioserver");

using AmazonS3Client client = new AmazonS3Client(config);

var resp = await client.GetObjectAsync("YourBucket", "YourObjectKey");
//do something with resp
```

The generic client I'm talking about is one that works for any S3-compatible endpoint, even when they have different input validation. The only reason there is a Google Cloud **and** an Amazon S3 client is because they differ in input validation. You'd probably just want to stick with the AmazonS3Client, but set the endpoint manually.

**Reply by Genbox** [reply]
I'm creating a working example for you, but it seems Minio is not playing along due to differences in the authorization header. I'll get back to you once I've tracked down what is going on.

**Comment by Genbox** [comment]
Alright. Here is a working sample:

```csharp
internal static class Program
{
    private static async Task Main()
    {
        AmazonS3Config config = new AmazonS3Config();
        config.EndpointTemplate = "{Scheme}://play.min.io/{Bucket}";
        config.Region = AmazonS3Region.UsEast1;
        config.Credentials = new StringAccessKey("QB8o9tNrpmTcFJSxs5aA", "2zmDVre0Sxp4bjlTUYQzVxfXhOmHe4NKWjoxKISG");
        config.NamingMode = NamingMode.PathStyle;

        using AmazonS3Client client = new AmazonS3Client(config);
        PutObjectResponse resp = await client.PutObjectStringAsync("testbucket", "objectname", "Hello World");
        Console.WriteLine("Success: " + resp.IsSuccess);
    }
}
```

This is using the minio playground. Buckets and access keys are removed regularly, so the sample might not work out-of-the-box when you run it.

If the access key or bucket is deleted, you can go to https://play.min.io and login with `minioadmin` as both the user and password. There you can create a new access key and re-create the bucket. Then the sample should work again.

**Comment by ivanjx** [comment]
thank you for your work @Genbox 
one more thing. is it necessary to specify `Region` for self hosted instances?

**Reply by Genbox** [reply]
Minio seems to default to us-east-1 no matter what region you choose. Region is required for the key signing process, so you just specify anything and it will work.
